src/sentry/replays/endpoints/project_replay_recording_segment_index.py

```diff
@@ -33,83 +33,119 @@ def get(self, request: Request, project, replay_id: str) -> Response:
             project_id=project.id,
             replay_id=replay_id,
         )
+        queryset_count = queryset.count()
 
-        if queryset.count() == 0:
-            return self.respond(status=404)
-
-        if request.GET.get("download") is not None:
+        if (
+            project.organization.id in settings.SENTRY_REPLAYS_DIRECT_FILESTORE_ORGS
+            and queryset_count == 0
+        ):
+            storage = get_storage(replays_storage_options())
+            # TODO: look at using threadpool to optimize download
             return self.paginate(
                 request=request,
-                queryset=queryset,
-                order_by="segment_id",
-                on_results=self.on_download_results,
+                on_results=functools.partial(download_files_direct, storage, project.id, replay_id),
                 response_cls=StreamingHttpResponse,
                 response_kwargs={"content_type": "application/json"},
-                paginator_cls=OffsetPaginator,
+                paginator_cls=GenericOffsetPaginator,
+                data_fn=functools.partial(get_files_direct, storage, project.id, replay_id),
             )
 
         else:
+            if queryset_count == 0:
+                return self.respond(status=404)
+
             return self.paginate(
                 request=request,
                 queryset=queryset,
                 order_by="segment_id",
-                on_results=lambda x: {
-                    "data": serialize(x, request.user, ReplayRecordingSegmentSerializer())
-                },
+                on_results=get_files_file_model,
+                response_cls=StreamingHttpResponse,
+                response_kwargs={"content_type": "application/json"},
                 paginator_cls=OffsetPaginator,
             )
 
-    def on_download_results(self, results):
-        """
-        get the files associated with the segment range requested. prefetch the files
-        in a threadpool.
-        """
-
-        # TODO: deflate files as theyre fetched, instead of having
-        # to wait untill all of them are complete before starting work.
-
-        recording_segment_files = File.objects.filter(
-            id__in=[r.file_id for r in results]
-        ).prefetch_related(
-            Prefetch(
-                "blobs",
-                queryset=FileBlobIndex.objects.select_related("blob").order_by("offset"),
-                to_attr="file_blob_indexes",
-            )
+
+def get_files_direct(storage, project_id, replay_id, offset, limit):
```

Comment:
probably need to write this function more precisely to handle pagination cases / add better unit tests